018-function-execution-model-refactor/018-function-execution-model-refactor.md

```diff
@@ -0,0 +1,111 @@
+
+## Problem Statement (Step 1)
+
+[problem-statement]: #problem-statement
+
+**What problem are you trying to solve?**
+
+Cloud Functions currently do not support the ability to execute synchronous functions and will require a complete rethink to how we approach how we execute our functions in order for us to realize this goal.
+
+**What is the context or background in which this problem exists?**
+
+Cloud Functions are currently a bit limited, this new design idea will allow for more customizability and control of how functions will execute and run for users.
+
+**Once the proposal is implemented, how will the system change?**
+
+- A new process will be introduced to execute functions
+- All current runtimes will be updated to use a web server as it's core
+- The Functions worker will be stripped of all direct docker interactions and will instead call the executor
+
+## Design proposal (Step 2)
+
+[design-proposal]: #design-proposal
+
+With this redesign of how cloud functions work the functions worker will no longer have to deal with spinning up docker containers but will instead use the executor server. Which will handle spinning up containers on behalf of the functions worker.
+
+The executor will have multiple tasks but the main one is to spin up runtimes when they are needed and to update their code whenever functions are updated. Runtimes will now be web servers based of their relevant languages and when code is added it will be added to the Runtime web server's router.
+
+The executor will also act like a router itself, when API Requests that are for functions are received it will be the executor's job to correlate them to the relevant runtimes and forward them for processing. When the Runtime's web server is done processing the request and returns a response the executor will return the response to whatever client requested it.
+
+The executor will be built using a Swoole HTTP Server and will be stored in the file: `app/executor.php`. The executor will need to be on the same network as appwrite so the two can communicate.
+
+#### API Endpoints
+This rewrite will add functionality to the API endpoints that are currently used by the functions worker. The executor will take over these endpoints while the rest will stay underneath the functions worker.
+
+`POST /v1/functions/{id}/executions/` - Execute function API endpoint
+
+The standard execution endpoint for functions will have functionality added for syncronous executions aswell as a new parameter in the JSON body to determine if this execution will be syncronous or asyncronous, this parameter will be called `isSync` and will be a boolean.
```

Comment:
I'm not sure that this param should be on the `POST /executions` route, but rather the `POST /tags` endpoint. Since sync and async serverless functions will be written differently (async scripts write to console log, sync scripts handle requests&responses), their execution models can't be interchanged. I think that we should store this fact in the same data structure as the rest of the tag info.

Reply:
Alright, that sounds like a good idea